### DISCUSSION TITLE
[FFmpeg-devel] [PATCH] lavc/nvenc: encode all RGB input as YUV444 rather than 420

### DISCUSSION THREAD

**Discussion by alatteri** [CESNET/UltraGrid#343] [General]
This looks useful.

http://ffmpeg.org/pipermail/ffmpeg-devel/2023-October/315276.html

**Comment by MartinPulec** [comment]
Well, IIRC when I've tried to add AV_PIX_FMT_X2RGB10 to the IS_YUV444() macro in FFmpeg, as it is in the patch, see comments [here](https://github.com/CESNET/UltraGrid/blob/b143ece377b6ad2afa63793115cff1a185e2c1bf/src/video_compress/libavcodec.cpp#L868-L871). I'll sum up:
1. without the patch, the x2rgb pixel format works correctly, but produces 10-bit YUV 4:2:0 only
2.  with the patch, the colors were not correct

for this reason the pixel format is disabled  - AV_PIX_FMT_GBRP16 is used for R10k and R12L input, which works correctly.

If you want to test after the patch gets upstream, you can try:

    uv -t testcard:codec=R10k -c lavc:enc=hevc_nvenc --param lavc-use-codec=x2rgb10

`lavc-use-codec` enforces the use of the _X2RGB_ pixfmt even though it is disabled otherwise. Anyways, as stated above, implicit _GBRP16_ works correctly producing RGB 4:4:4 10b stream now. Of course I'd be happy if we won't need to disable the pixfmt - it could have slight performance benefits (less RAW data) and there won't be exceptions in the code required. 

I didn't test 8-bit RGB formats in the past, just now, and without the patch, point 1 holds (8b YUV 4:2:0 is encoded).
